### DISCUSSION TITLE
Micronaut 3 to 4 migration issue: request.getHaaders.getFirst (key, type.class) can no longer convert

### DISCUSSION THREAD

**Discussion by ethanhere** [micronaut-projects/micronaut-core#10899] [General]
In micronaut 3 we have the following code working
	io.micronaut.http.HttpRequest<?> request = ...
        HttpHeaders headers = request.getHeaders();
        Optional<MyAppId> optionalMyAppId = headers.getFirst("id",
            MyAppId.class);
        final MyAppId appId = optionalMyAppId.orElse(null);

Where MyAppId is a pojo class:
public final class MyAppId {
    private final String value;
    private MyAppId(String value) {
        this.value = value;
    }

    @JsonCreator
    public static MyAppId valueOf(@JsonProperty String value) {
        MyAppIdValidator.validateMyAppId(value);
        return new MyAppId(value);
    }

    @JsonCreator
    public static AccountId of(@JsonProperty String value) {
        return valueOf(value);
    }
}

In micronaut 3, it will get the value of "id" header and use MyAppId.of method to create a MyAppId object.

This code no longer works in micronaut 4. I debugged into the code and found in
https://github.com/micronaut-projects/micronaut-core/blob/0118db96962eaf7a83abbcb81c1afd73139ff777/core/src/main/java/io/micronaut/core/convert/DefaultMutableConversionService.java#L237
It will return EmptyAnnotationMetadata and not able to use of method to convert.

Can you advise how to make our original code working?

**Comment by sdelamo** [comment]
Please, attach a sample application using Micronaut Framework 3 working and the same 4 with Micronaut Framework 4 showing the regression. @dstepanov any clue?

**Reply by ethanhere** [reply]
I debugged in micronaut 4 environment, headers.getFirst("id",
MyAppId.class) will get into DefaultMutableConversionService.convert(S object, Class<? super S> sourceType, Class<T> targetType, ConversionContext context)
Then it won't convert.

The first question is: what's the point of method ConvertibleValues.getFirst(CharSequence name, Class<T> requiredType) ?
HttpHeaders is a sub class of ConvertibleValues.

My understand of ConvertibleValues.getFirst(CharSequence name, Class<T> requiredType) is to return header value of the name in requiredType object. Otherwise, ConvertibleValues.getFirst(CharSequence name) is enough, correct?

So how can I get ConvertibleValues.getFirst(CharSequence name, Class<T> requiredType) working in micronaut 4?

**Reply by graemerocher** [reply]
attach an example to an issue report

**Reply by ethanhere** [reply]
[githubrepro.zip](https://github.com/user-attachments/files/15839433/githubrepro.zip)
Hi @graemerocher @sdelamo @diroussel @aalmiray attached is a repro. Please take a look and suggest a fix. Thanks.

**Reply by graemerocher** [reply]
could you create an associated issue and attach it there. Thanks.

**Reply by ethanhere** [reply]
[githubrepro.zip](https://github.com/user-attachments/files/15845171/githubrepro.zip)
attaching the reproducer. 
What do you mean by associated issue? Can you see the attached zip file? you can create that issue and put the reproducer file there. @graemerocher

**Reply by sdelamo** [reply]
I have converted this into an issue. https://github.com/micronaut-projects/micronaut-core/issues/10914
